File: src/ocaml/renderer/websocket_handler.ml
(** Providing functions that handle message of Websocket.   *)
type t = {
  ws: WebSockets.webSocket Js.t;
  mutable handlers: (WebSockets.webSocket WebSockets.messageEvent Js.t -> unit) list;
}

let make ws = {
  ws;
  handlers = [];
}

let init t =
  t.ws##.onmessage := Dom.handler (fun message ->
      List.iter (fun handler -> handler message) t.handlers;
      Js._true)

let add t ~handler = t.handlers <- handler :: t.handlers

File: src/ts/libs/json-rpc/websocket-handler.ts
// handler definition
export type Handler = {
  handle(ev: MessageEvent): void;
};

// This module defines handler of WebSocket.
export class WebSocketHandler {
  private handlers: Handler[];
  constructor(private ws: WebSocket) {
    this.handlers = [];
  }

  /**
   *  initialize message handler for WebSocket.
   */
  public initialize() {
    this.ws.onmessage = ev => this.handlers.forEach(handler => handler.handle(ev));
  }

  /**
   *  Add handler to this class.
   *
   *  @param handler a function to handle message.
   */
  public addHandler(handler: Handler) {
    this.handlers.push(handler);
  }
}
